src/Feature/Theme/Feature_Theme.re

```diff
@@ -45,34 +49,50 @@ let initial = contributions => {
   theme: ColorTheme.{variant: Dark, colors: ColorTheme.Colors.empty},
 };
 
-let resolver =
+let colors =
     (
       ~extensionDefaults as _=[], // TODO
       ~customizations=ColorTheme.Colors.empty, // TODO
       model,
     ) => {
   let {schema, theme} = model;
 
-  let resolve = key => {
+  let rec resolve = key => {
     switch (ColorTheme.Colors.get(key, customizations)) {
-    | Some(color) => `Color(color)
+    | Some(color) => Some(color)
     | None =>
       switch (ColorTheme.Colors.get(key, theme.colors)) {
-      | Some(color) => `Color(color)
+      | Some(color) => Some(color)
       | None =>
         switch (ColorTheme.Schema.get(key, schema)) {
         | Some(definition) =>
-          `Default(
-            ColorTheme.Defaults.get(theme.variant, definition.defaults),
-          )
+          definition.defaults
+          |> ColorTheme.Defaults.get(theme.variant)
+          |> ColorTheme.Defaults.evaluate(resolve)
 
-        | None => `NotRegistered
+        | None => None
         }
       }
     };
   };
 
-  resolve;
+  let defaults =
+    schema
+    |> ColorTheme.Schema.toList
+    |> List.map(
+         ColorTheme.Schema.(
+           definition => (
+             definition.key,
+             definition.defaults
+             |> ColorTheme.Defaults.get(theme.variant)
+             |> ColorTheme.Defaults.evaluate(resolve)
+             |> Option.value(~default=Color.hex("#0000")),
+           )
+         ),
+       )
+    |> ColorTheme.Colors.fromList;
+
+  ColorTheme.Colors.unionMany([defaults, theme.colors, customizations]);
 };
```

Comment:
This is the flattening, which is a bit awkward because it now duplicates the fallback logic. Perhaps it would be better to just get the keys from the schema and resolve every value, instead of just the default fallbacks.